### PR TITLE
#568 event auditor

### PR DESCRIPTION
**Pull request by joncppl**
Notes:

-used a bit of component state in EventAuditor/index.js, mostly to keep the count of is_submission events correct without fetching the list again / writing a custom reducer.

### CODE REVIEW

File: dispatch/apps/api/views.py [hunk 1]
```diff
@@ -403,6 +404,10 @@ def get_queryset(self):
                 Q(description__icontains=q) |
                 Q(host__icontains=q)
             )
+        if pending:
+            queryset = queryset.filter(
+                Q(is_submission__exact=1)
```
Comment (psiemens) [inline dispatch/apps/api/views.py]: You don't need to use `Q` here since there's only one condition. Also, the default queryset should exclude submitted events:

```py
if pending:
    queryset = queryset.filter(is_submission=True)
else:
    queryset = queryset.filter(is_submission=False)
```

Can you add a test case for this as well?

File: dispatch/static/manager/src/js/components/Header.js [hunk 2]
```diff
@@ -54,6 +54,9 @@ export default function Header(props) {
                 <li className='c-header__dropdown__link'>
                   <Link to='/events/'><span className='pt-icon-standard pt-icon-timeline-events' /> events</Link>
                 </li>
+                <li className='c-header__dropdown__link'>
+                  <Link to='/events/audit'><span className='pt-icon-standard pt-icon-endorsed' /> audit events</Link>
+                </li>
```
Comment (psiemens) [inline dispatch/static/manager/src/js/components/Header.js]: Instead of adding a separate link in the navigation, I think it would be better to link to this page from the main events listing. Something like this:

screen shot 2017-06-13 at 9 16 06 am

Clicking on "5 pending events" would take the user to the approval page.

I used the Blueprint tag component with the following classNames:

```html
<div class="pt-tag pt-large pt-intent-warning">5 pending events</div>
```

File: dispatch/static/manager/src/js/components/inputs/LinkButton.js [hunk 3]
```diff
@@ -5,7 +5,12 @@ import { Button } from '@blueprintjs/core'
 export default function LinkButton(props) {
   return (
     <Link
-      to={props.to}>
+      to={props.to}
+      onClick={e => {
+        if (props.disabled) {
+          e.preventDefault()
```
Comment (psiemens) [inline dispatch/static/manager/src/js/components/inputs/LinkButton.js]: Nice addition 👍

File: dispatch/static/manager/src/js/index.js [hunk 4]
```diff
@@ -100,6 +100,7 @@ render((
           <Route path='events'>
             <IndexRoute component={Pages.Events.Index} />
             <Route path='new' component={Pages.Events.NewEvent} />
+            <Route path='audit' component={Pages.EventAudit} />
```
Comment (psiemens) [inline dispatch/static/manager/src/js/index.js]: The audit page should be grouped with the other events pages and referenced like so:

`component={Pages.Events.Audit}`

File: dispatch/static/manager/src/js/pages/EventAudit/EventAuditPagination.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+export default function EventAuditPagination(props) {
+  const { page, pages } = props
+
+  return (
+    <div className='c-event-audit-pagination'>
+      <LinkButton
+        to={getPath(props.location, page - 1)}
+        disabled={page <= 1}>Prev</LinkButton>
+      {' '}
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/EventAuditPagination.js]: Use `margin` to add space between buttons rather than a hardcoded space

File: dispatch/static/manager/src/js/pages/EventAudit/EventAuditPagination.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+function getPath(location, page) {
+  let query = R.clone(location.query)
+
+  query.page = page
+
+  return {
+    pathname: location.pathname,
+    query: query
+  }
+}
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/EventAuditPagination.js]: Is this done anywhere else in the application? Does it make sense to move to `util/helpers.js`?

File: dispatch/static/manager/src/js/pages/EventAudit/EventCard.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import React from 'react'
+
+import { AnchorButton, Tooltip } from '@blueprintjs/core'
+import { LinkButton } from '../../components/inputs'
+import { humanizeDatetime } from '../../util/helpers'
+
+function labelAndProp(label, prop) {
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/EventCard.js]: Convert this to a functional React component, then call like so below:

`<EventLabel label='Description' value={this.props.event.description} />`

File: dispatch/static/manager/src/js/pages/EventAudit/EventCard.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+        <div className='c-event-audit-card-button-border' />
+        <div className='c-event-audit-card-button-row'>
+
+          <Tooltip
+            className='c-event-audit-card-button-right'
+            transitionDuration={15}
+            content="Delete">
+            <AnchorButton
+              disabled={this.state.disableButtons}
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/EventCard.js]: Can you explain the need for `this.state.disableButtons`?

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+  pending: 1,
+  limit: PER_PAGE
+}
+
+class EventAuditPage extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      count: 0
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: How hard would it be to control this in the store? I know it works this way, but it breaks out of the redux architecture and seems kinda hard to follow. I might be missing something though

Ideally the events list that we keep in the store would stay up to date, and any time an event goes from `is_submission=True` to `is_submission=False`, that change would be reflected. Then you can simply count up all of the events with `is_submission=True` and use that as your count

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    }
+  }
+
+  getQuery(props) {
+    props = props || this.props
+
+    const query = R.clone(PENDING_QUERY)
+
+    if (props.location.query.page) {
+      query.offset = (Number(props.location.query.page) - 1) * PER_PAGE
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: use `parseInt(props.location.query.page, 10)` instead of `Number`

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  approve(id) {
+    const event = this.props.entities.events[id]
+    event.is_submission = 0
+    this.props.saveEvent(this.props.token, id, event)
+    this.checkPage()
+    this.setState({ count: this.state.count - 1 })
+    this.reloadListFlag = false
+  }
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: I think it might be better to create a new action for this, and then just call `this.props.approveEvent(this.props.token, id)`

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+  disapprove(id) {
+    this.props.deleteEvent(this.props.token, id)
+    this.checkPage()
+    this.setState({ count: this.state.count - 1 })
+    this.reloadListFlag = false
+  }
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: Same for disapproving events

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+  render() {
+    this.eventCount = 0
+    const events = this.props.events.ids.map((id, idx) => {
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: I prefer to chain this logic together like so:

```js
const events = this.props.events.ids
  .map(id => this.props.entities.events[id])
  .map((event, i) => {
    // ...
  })

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+      const event = this.props.entities.events[id]
+
+      if (!event.is_submission) {
+        return null
+      }
+      this.eventCount++
+
+      return (
+        <EventCard
+          key={idx}
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: Use `i` instead of `idx` (for consistency with rest of application)

File: dispatch/static/manager/src/js/pages/EventAudit/index.js [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+          </h2>
+          <div className='c-event-audit-subtitle'>
+            {this.state.count > PER_PAGE ?
+              <EventAuditPagination
+                page={Number(this.props.location.query.page || 1)}
+                pages={Math.ceil(this.state.count / PER_PAGE)}
+                location={this.props.location} />: null}
+            {this.state.count} event{this.state.count == 1 ? '' : 's'} pending approval
+          </div>
+          <ReactCSSTransitionGroup
```
Comment (psiemens) [inline dispatch/static/manager/src/js/pages/EventAudit/index.js]: This transitions are nice, but I think it'd prefer to remove them for now since we don't use them anywhere else in the application

**Review by psiemens — CHANGES_REQUESTED** [review]
This is really good overall. I think we might disagree on some of the code design choices, so feel free to argue your side.

I try to manage the applications state entirely from inside the store so that components deal with as little "business logic" as possible. There are times when that pattern should be broken, but I'm not sure this is one of them.

File: dispatch/static/manager/src/js/pages/EventAudit/EventCard.js (hunk 8 above)
Comment (joncppl) [reply dispatch/static/manager/src/js/pages/EventAudit/EventCard.js]: just to disable them during the animation.
Since I'll be removing that it won't be necessary anymore

File: dispatch/apps/api/views.py (hunk 1 above)
Comment (joncppl) [reply dispatch/apps/api/views.py]: looks like axel has a test in Event-forms-etc. 
When that's merged I'll pull it into this branch.

Comment (Axel-Jacobsen) [reply dispatch/apps/api/views.py]: I completed this in my PR because we were both modifying the same thing. Feel free to modify as needed!

Comment (psiemens) [reply dispatch/apps/api/views.py]: Sweet! my bad -- didn't realize at first

**Review by psiemens — APPROVED** [review]
